### DISCUSSION TITLE
Does the plugin API version always match the Composer version?

### DISCUSSION THREAD

**Discussion by mbrodala** [composer/composer#10030] [Q&A / Support] [answered]
Based on https://github.com/renovatebot/renovate/issues/11019:

Composer automatically updates the `plugin-api-version` in its lockfile whenever any package update is performed. If that is done by a newer Composer version, the `plugin-api-version` is bumped. So e.g. if the lockfile was last updated by Composer 2.0.x and Composer has been updated to version 2.1.x in the meantime, the next update of any arbitrary package will bump the `plugin-api-version` to `2.1.0`.

So does the plugin API version always match the Composer version?

Or phrased differently given the context of the mentioned ticket: can one safely determine the last Composer major+minor version which was used to update the lockfile?

The change mentioned in the ticket (https://github.com/composer/composer/commit/7dd95a768bc1dd8c6f296ee0081f41ffa1444bbe) implies this slightly since this plugin API version bump was published with the first RC of Composer 2.1.

**Comment by Seldaek** [comment, marked as answer]
In short.. no. It only gets bumped when we do add features, so most likely it'll only at best match the major.minor version of the composer used. But if nothing changes for plugins most likely we will not even bump it, and it may remain at 2.1.0 for a while.

I am not sure what you are trying to achieve here though?

**Reply by mbrodala** [reply]
Thanks for the insights. That the plugin API version matches the Composer version in major/minor would be good. But that it doesn't necessarily change on Composer version change does not match my expectation but makes sense.

My goal is teaching Renovate (see linked issue) to keep using the Composer version which was last used to update the lockfile. This way I can avoid unintentional package updates.

Scenario: we have various dependencies including packages which are Composer plugins. If one plugin has an update available which requires a higher plugin API version it could be updated together with the actual updates. Now if this is not reviewed properly it could get merged which would lead to errors when doing install/update with older Composer versions.

Not sure if there is a unwritten rule that Composer plugin packages must do a major version bump if they require a newer Composer plugin API. If yes, then we can probably ignore plugin API version bumps in the lockfile since major updates of any package must be done intentionally and separately anyways.

**Reply by Seldaek** [reply]
There is no such unwritten rule. IMO dependency bumps are fine in minor versions as the dependency resolution will prevent bad updates.

So yeah, I guess you could make sure to run with composer matching plugin-api-version in terms of major.minor, but take the last patch release at least.

Ideally you would take the highest composer version which still has the same plugin-api-version though, as otherwise you risk using a very old version at some point if we do not bump the plugin api for a while.

**Reply by mbrodala** [reply]
The last point is indeed important. We'd be stuck with Composer 2.1.x forever if the plugin API version is not changed in Composer 2.2.x and newer.

I assume that the runtime API version (`composer-runtime-api`) is also only updated on real API change, right?

So technically I cannot see a reliable way do determine major+minor version of the last Composer version which was used for a lockfile update. For me this means I'd need a way to configure this in Renovate. This way I can ensure that Renovate uses the same Composer version as I.

**Reply by Seldaek** [reply]
runtime-api is the same deal yes. We'll introduce a new `composer` platform package in 2.2 (https://github.com/composer/composer/issues/9963) but this isn't helping as it won't be in the lock file (which would be a pita IMO as minor version changes would keep changing the lock file on every update/require you do even if nothing else changes).

I also don't really see a perfect way out of this, sorry :)

**Reply by mbrodala** [reply]
No problem, you already helped me a lot. As mentioned I'll try to get this solved in Renovate itself.

**Reply by mbrodala** [reply]
Also https://github.com/composer/composer/issues/9963 sounds like it could actually help with my issue. We could then require e.g. `"composer": "~2.1.0"` and be sure that only Composer 2.1.x is used. :thinking:

**Reply by Seldaek** [reply]
Yeah that could be an option, although it's most likely to cause friction in the team and be quite annoying.. I wonder if it's really worse than dealing with the occasional bad update with the current setup, which I imagine would be rather infrequent.

**Reply by mbrodala** [reply]
You are probably right. I guess we'll just leave it as it is and see if we really encounter issues. We don't plan to lag behind Composer releases in our dev environment for too long anyways.
